Bad-ptr/persp-mode.el
# Set up automatic testing

## How to set up automatic testing

This PR does most of the work for setting up automatic tests with [Travis CI](https://travis-ci.org/), including coverage report with [coveralls.io](https://coveralls.io/). Before merging this, you should:
- sign up to Travis and hook Travis to persp-mode.el, according to steps 1 and 2 in https://docs.travis-ci.com/user/getting-started/ (steps 3 and 4 are taken care of by this PR)
- sign up to Coveralls and add persp-mode.el to Coveralls

After merging this PR, you should be able to view test results in Travis and coverage report in Coveralls. For example, here are the [test results for my branch](https://travis-ci.org/bmag/persp-mode.el). For some reason, I couldn't get [undercover.el](https://github.com/sviridov/undercover.el) to send coverage reports to Coveralls. I hope it is something local to my fork, and that it will just work when you merge. Even if it doesn't, testing with Travis will still work, you just won't have a coverage report.

I didn't squash the commits, because I wanted them to match the build history in my example link above.

I think you might prefer to merge the PR into a temporary branch, instead of directly to master, but that's up to you.

## How to use this

Every time you push changes to Github, a Travis build will start automatically. When the build finishes, you can view the results in Travis. You can also set it up to send you emails when tests fail (that might be the default behavior, actually).

To run the tests locally, you will have to install [Cask](https://github.com/cask/cask). Running the tests is simple as:
```sh
# install buttercup and undercover (the dependencies specified in Cask file)
make install
# run the tests
make test
```
If you have several Emacs versions installed, you can run the tests for a specific version easily:
```sh
EMACS=/path/to/emacs make install
EMACS=/path/to/emacs make test
```

## Writing tests

Tests are written with [Buttercup](https://github.com/jorgenschaefer/emacs-buttercup), and instructions are [here](https://github.com/jorgenschaefer/emacs-buttercup/blob/master/docs/writing-tests.md).

I tried to write some basic tests that enable `persp-mode` and then test `persp-switch`, but for some reason they had errors and failed. I think maybe I didn't enable `persp-mode` correctly (I just called `(persp-mode)`), but didn't investigate.
